### PR TITLE
Lifetime profit

### PR DESCRIPTION
**Pull request by SamuXarick**
There is now a lifetime profit value against vehicles which can be viewed alongside the profit this year and profit last year.

burty - Stores lifetime profit (gets updated at end of year)
burty - Displays on the vehicle info window where the other profits are shown.
burty - Saves the lifetime profit in to the save game and loads it.
burty - Sort by lifetime profit in vehicle lists.
JGR   - Display the current lifetime profit, instead of the lifetime profit at the start of the current year
JGR   - Reset overall lifetime profit to 0 when renewing vehicle.
JGR   - Sort by length was incorrectly available, and crashed for non ground vehicles.
Samu  - Displays on the vehicle group info panel where the other profits are shown.
Samu  - Displays on the vehicle list window where the other profits are shown.
Samu  - Add AI/GS functions to get lifetime profits of groups and vehicles.

### CODE REVIEW

File: src/lang/english.txt [hunk 1]
```diff
@@ -3436,7 +3437,7 @@ STR_VEHICLE_LIST_ROAD_VEHICLE_TOOLTIP                           :{BLACK}Road veh
 STR_VEHICLE_LIST_SHIP_TOOLTIP                                   :{BLACK}Ships - click on ship for information
 STR_VEHICLE_LIST_AIRCRAFT_TOOLTIP                               :{BLACK}Aircraft - click on aircraft for information
 
-STR_VEHICLE_LIST_PROFIT_THIS_YEAR_LAST_YEAR                     :{TINY_FONT}{BLACK}Profit this year: {CURRENCY_LONG} (last year: {CURRENCY_LONG})
+STR_VEHICLE_LIST_PROFIT_THIS_YEAR_LAST_YEAR_LIFETIME            :{TINY_FONT}{BLACK}Profit this year: {CURRENCY_LONG} (last year: {CURRENCY_LONG}) (lifetime: {CURRENCY_LONG})
```
Comment (nielsmh) [inline src/lang/english.txt]: I don't think there is a precedent for this, but I'd prefer it formatted like `(last year: £xx, lifetime: £xx)`, so not two parentheses after each other, but comma separated in a single parenthesis.

File: src/saveload/saveload.h [hunk 2]
```diff
@@ -302,6 +302,7 @@ enum SaveLoadVersion : uint16 {
 	SLV_MULTITILE_DOCKS,                    ///< 216  PR#7380 Multiple docks per station.
 	SLV_TRADING_AGE,                        ///< 217  PR#7780 Configurable company trading age.
 	SLV_ENDING_YEAR,                        ///< 218  PR#7747 Configurable ending year.
+	SLV_LIFETIME_PROFIT,                    ///< 219          Lifetime profit.
```
Comment (nielsmh) [inline src/saveload/saveload.h]: Add the PR number :)

File: src/saveload/vehicle_sl.cpp [hunk 3]
```diff
@@ -691,6 +691,7 @@ const SaveLoad *GetVehicleDescription(VehicleType vt)
 		 SLE_CONDVAR(Vehicle, profit_this_year,      SLE_INT64,                   SLV_65, SL_MAX_VERSION),
 		 SLE_CONDVAR(Vehicle, profit_last_year,      SLE_FILE_I32 | SLE_VAR_I64,   SL_MIN_VERSION,  SLV_65),
 		 SLE_CONDVAR(Vehicle, profit_last_year,      SLE_INT64,                   SLV_65, SL_MAX_VERSION),
+		 SLE_CONDVAR(Vehicle, profit_lifetime,       SLE_INT64,                  SLV_LIFETIME_PROFIT, SL_MAX_VERSION),
```
Comment (nielsmh) [inline src/saveload/vehicle_sl.cpp]: Alignment looks slightly wrong here, seems to be missing a space before SLV_LIFETIME_PROFIT.

File: src/script/api/script_group.hpp [hunk 4]
```diff
@@ -205,6 +205,14 @@ class ScriptGroup : public ScriptObject {
 	 */
 	static Money GetProfitLastYear(GroupID group_id);
 
+	/**
+	 * Get the lifetime profit of a group.
+	 * @param group_id The group to get the profit of.
+	 * @pre IsValidGroup(group_id).
+	 * @return The current lifetime profit the group has.
+	 */
+	static Money GetLifetimeProfit(GroupID group_id);
```
Comment (nielsmh) [inline src/script/api/script_group.hpp]: `GetProfitLifetime`? To match the pattern of GetProfitThisYear and GetProfitLastYear.

**Review by nielsmh — COMMENTED** [review]
Most of these comments are just thoughts and opinions.

Before reading this code I kind of expected the lifetime profit to include the purchase cost of the vehicle/train. On the other hand, that might be difficult to keep track of with trains composed of multiple vehicles, i.e. what happens if you add or remove cars from a train, how does that update the lifetime profit? (I think that can be solved, but I'm not sure if it's worth solving.)

File: src/saveload/vehicle_sl.cpp (hunk 3 above)
Comment (LordAro) [reply src/saveload/vehicle_sl.cpp]: Could fix profit_last_year a couple of lines above as well, while you're in the area

File: src/script/api/script_group.hpp (hunk 4 above)
Comment (LordAro) [reply src/script/api/script_group.hpp]: I'm not sure here. It's slightly more consistent perhaps, but it doesn't read as well.
"get the profit from last year" -> GetProfitLastYear
"get the lifetime profit" -> GetLifetimeProfit
"get the profit from lifetime" reads very weirdly to me

**Comment by SamuXarick** [comment]
I have a dilema - what should be the conversion approach for old savegames?

Renewed vehicles reset lifetime profit to 0, but not last year profit. It remains whatever it is.
This makes it so that vehicles before 1 year old of age from old savegames to have uncertain last year profit.

It could be a real new vehicle, or a renewed vehicle. If it's a real new vehicle, then getting its last year profit as best guess for lifetime profit is correct, no matter its age. But if the vehicle was renewed, then its last year profit was inherited from its past vehicle, and using its last year profit to determine lifetime profit, falls into the wrong side, unless the vehicle is already 1 year old at the time of conversion.

How should I do the lifetime conversion?
- As burty had in his patch, profit_lifetime equals to 0 for every vehicle.
- As it currently stands, best guess by using last years profit, no matter the age.
- Using last years profit for vehicles that are already 1 year old, but 0 for the others.

**Comment by SamuXarick** [comment]
Another problem: Too large profits don't fit.
screenshot113 (8)

**Comment by SamuXarick** [comment]
Better now.
Unnamed, 2036-10-18

**Comment by stormcone** [comment]
Wouldn't it be better to make the window resizable horizontally too or to put a horizontal scrollbar at the bottom of the vehicle list? Instead of making wider the whole window by default.

**Comment by TrueBrain** [comment]
This is once again a Pull Request which is impossible for us to review. It contains thoughts, it contains experiments, it contains lines of code that do not appear to belong in this PR. I am so confused, I am not even willing to look at this anymore.

Please Samu, start with small PRs, of like 1 function, 1 tiny change. Look how other people are doing it, and how much more smooth we can guide them through this process. This PR is all over the place, making it impossible to guide you in any way that will lead to a result. And this is not the first time we are telling you this .. in fact, your recent PRs show that you are not learning at all (and there is a year between those two!).

So again, I am not against this idea, but we do not have the time to go through this implementation and ensure it does not contain any game-breaking side-effects.

**Pull request closed without merging**